crates/doc/src/inference.rs

```diff
@@ -1484,6 +1486,201 @@ impl Shape {
     }
 }
 
+const MAX_ROOT_FIELDS: usize = 750;
+const MAX_NESTED_FIELDS: usize = 200;
+
+impl Shape {
+    // Widen a Shape to make the provided AsNode value fit.
+    // Returns a hint if some locations might exceed their maximum allowable size.
+    // NOTE: If a particular location defines `additionalProperties` as a subschema, don't
+    // add the field from `AsNode`, instead jump straight to squashing the the field into `additionalProperties`
+    fn widen<'n, N>(&mut self, node: &'n N) -> bool
+    where
+        N: AsNode,
+    {
+        self.widen_inner(node, Location::Root)
+    }
+
+    fn widen_inner<'n, N>(&mut self, node: &'n N, loc: Location) -> bool
+    where
+        N: AsNode,
+    {
+        use crate::{Field, Fields};
+
+        match node.as_node() {
+            crate::Node::Object(obj) => {
+                self.type_ = self.type_ | types::OBJECT;
+
+                match self.object.additional {
+                    Some(_) => {}
+                    None => {
+                        self.object.additional = Some(Box::new(Shape {
+                            type_: types::INVALID,
+                            ..Default::default()
+                        }))
+                    }
+                }
+
+                // If a particular location defines `additionalProperties` as a subschema
+                // don't add the field from `AsNode`, instead jump straight to squashing
+                let hint = match self.object.additional.as_mut() {
+                    // We only want to squash if your `additionalProperties`
+                    // is set to something other than `false`.
+                    Some(addl) if !addl.type_.eq(&types::INVALID) => {
+                        obj.iter().enumerate().fold(false, |accum, (idx, node)| {
+                            accum || addl.widen_inner(node.value(), loc.push_item(idx))
```

Comment:
this should be `loc.push_property`. `push_item` is only for arrays.

Reply:
This is in the `Node::Array` match arm, so I figured `push_item` was correct since it took a numeric index. But since we're not actually widening tuple shapes, I'm realizing now this index should probably always be `0`... or something? It's not a huge deal because the Location is only used at present to detect if we're at `Location::Root` or not, but if we have it I'd like it to be as correct as possible